src/Hazelcast.Net/HazelcastClient.cs

```diff
@@ -79,51 +79,88 @@ public HazelcastClient(HazelcastOptions options, Cluster cluster, ILoggerFactory
+
+            // when a connection is closed, notify the members service (may change the client state)
+            Cluster.Connections.ConnectionClosed += async conn => { await Cluster.Members.RemoveConnectionAsync(conn).CfAwait(); };
 
             // when a connection is closed, trigger the user-level event
-            Cluster.Connections.ConnectionClosed
-                += conn => Trigger<ConnectionClosedEventHandler, ConnectionClosedEventArgs>(new ConnectionClosedEventArgs(conn));
-
-            // when a connection is opened, DistributedObjects.OnConnectionOpened checks whether it is the first
-            // connection to a new cluster, and then re-creates all the known distributed object so far on the
-            // new cluster. it does so by using the new connection, which should be stable: the only reason why
-            // a new connection could be teared down by ClusterMembers is if there already are known connections,
-            // which cannot be the case if this is the first connection. So we are not going to, like, jump to
-            // another connection while this runs - there's one and only connection for now.
-            Cluster.Connections.ConnectionOpened
-                += _distributedOjects.OnConnectionOpened;
+            Cluster.Connections.ConnectionClosed += conn => Trigger<ConnectionClosedEventHandler, ConnectionClosedEventArgs>(new ConnectionClosedEventArgs(conn));
+
+            // -- ConnectionOpened -- order is *important* --
+
+            // when the first connection to a cluster is opened, distributed objects may be re-created, compact
+            // schemas may be published, etc. this is all performed using the new connection, which should be
+            // stable. if it dies, then the client ends up disconnected again, and everything starts from scratch.
+            //
+            // while attempting the first connection, there is no members-connection task running, so there cannot
+            // be any other connection: it is not possible that this new connection dies and the client ends up
+            // running with other connections, but not fully initialized.
+
+            // when the first connection to a cluster is opened, make sure we recreate the distributed objects
+            Cluster.Connections.ConnectionOpened += _distributedOjects.OnConnectionOpened;
 
             // when the first connection to a cluster is opened, make sure we republish all schemas
-            Cluster.Connections.ConnectionOpened
-                += Cluster.SerializationService.CompactSerializer.Schemas.OnConnectionOpened;
+            Cluster.Connections.ConnectionOpened += Cluster.SerializationService.CompactSerializer.Schemas.OnConnectionOpened;
+
+            // when a connection is opened, install event subscriptions + ensure there is a cluster views connection
+            Cluster.Connections.ConnectionOpened += Cluster.Events.OnConnectionOpened;
```

Comment:
May be it's a silly question but how is it likely that a subscription is triggered before client state is set to "connected"?

Reply:
as soon as the subscription is registered on the server, we can receive an event ... and the client may end up handling that event *before* it is considered "connected" which might indeed be considered a Bad Idea. 

but if we do it the other way round, the client may do a map.put operation as soon as connected, assuming it will trigger and event and *not* get an event because the subscription hasn't been registered yet.

java does subscriptions as "urgent" invocations - so they'll be registered before the client is fully connected, too. so I *think* we're OK.